### PR TITLE
Fix several tests with resource leakage

### LINKED ISSUE #1226: Checking Resource Leak at end of a test class
Recently unit test can not complete due to threads leaking and ZK session leaking. We identified the leaks and fixed them. In order for future test not to leak resources, we need a way to enforce test class would release/clean up resource created during test.

A check can/should be implemented in @afterclass in ZkTestBase.java.

### PR DESCRIPTION
**Pull request by kaisun2000**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

fix #1226

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

Fix several tests with resource leakage

### Tests

- [ ] The following tests are written for this issue:

(List the names of added unit/integration tests)

- [x] The following is the result of the "mvn test" command on the appropriate module:
github run pending

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java [hunk 1]
```diff
@@ -72,36 +73,45 @@ public void testController() throws Exception {
...
+
+      // path = PropertyPathConfig.getPath(PropertyType.LEADER, clusterName);
+      // ZNRecord leaderRecord = _gZkClient.<ZNRecord> readData(path);
+      LiveInstance liveInstance = accessor.getProperty(keyBuilder.controllerLeader());
+      AssertJUnit.assertEquals(controllerName, liveInstance.getInstanceName());
+      // AssertJUnit.assertNotNull(election.getController());
+      // AssertJUnit.assertNull(election.getLeader());
+    } finally {
+      manager.disconnect();
+      controller0.shutdown();
```
Comment (jiajunwang) [inline helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java:88]: nit, shutdown the controller first would be safer.

File: helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java [hunk 2]
```diff
@@ -72,36 +73,45 @@ public void testController() throws Exception {
...
+    try {
+      election.onControllerChange(context);
+      // leaderRecord = _gZkClient.<ZNRecord> readData(path);
+      LiveInstance liveInstance = accessor.getProperty(keyBuilder.controllerLeader());
+      AssertJUnit.assertEquals(controllerName, liveInstance.getInstanceName());
+      // AssertJUnit.assertNull(election.getController());
+      // AssertJUnit.assertNull(election.getLeader());
+    } finally {
+      manager.disconnect();
+      controller1.shutdown();
```
Comment (jiajunwang) [inline helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java:106]: Same here.

File: helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java [hunk 3]
```diff
@@ -197,35 +218,42 @@ public void testCompeteLeadership() throws Exception {
...
+              return false;
+            }
           }
-        }
-      }, 1000));
+        }, 1000));
+      }
+    } finally {
+      for (GenericHelixController controller : controllers) {
+        controller.shutdown();
```
Comment (jiajunwang) [inline helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java:254]: How about the managers? Please disconnect them too.

File: helix-core/src/test/java/org/apache/helix/tools/TestClusterSetup.java [hunk 4]
```diff
@@ -73,44 +73,63 @@
...
 
-  @Test
+  // Note, with mvn 3.6.1, we have a nasty bug that running "mvn test" under helix-core,
+  // all the bellow test will be invoked after other test including @AfterClass cleanup of this
+  // This bug does not happen of running command as "mvn test -Dtest=TestClusterSetup". Nor does it
+  // happen in intellij. The workaround found is to add dependsOnMethods attribute to all the rest.
+  @Test(dependsOnMethods = "testAddClusterWithValidCloudConfig")
   public void testZkAdminTimeout() {
     boolean exceptionThrown = false;
+    ZKHelixAdmin admin = null;
```
Comment (jiajunwang) [inline helix-core/src/test/java/org/apache/helix/tools/TestClusterSetup.java:103]: nit, you probably want to extract a private method for this section to reduce duplication.

File: helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java (hunk 3 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java:254]: good point. added.

File: helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java (hunk 1 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java:88]: changed.

File: helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java (hunk 2 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java:106]: changed.

File: helix-core/src/test/java/org/apache/helix/tools/TestClusterSetup.java (hunk 4 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/tools/TestClusterSetup.java:103]: added an test helper to reduce duplicated code.

File: helix-core/src/test/java/org/apache/helix/controller/changedetector/TestResourceChangeDetector.java [hunk 5]
```diff
@@ -431,8 +431,12 @@ public void testResetSnapshots() {
             .setResources(new HashSet<>(_dataAccessor.getChildNames(_keyBuilder.idealStates())))
             .setWaitTillVerify(TestHelper.DEFAULT_REBALANCE_PROCESSING_WAIT_TIME)
             .build();
-    Assert.assertTrue(_clusterVerifier.verify());
-
+    try {
+      boolean result = _clusterVerifier.verify();
```
Comment (xyuanlu) [inline helix-core/src/test/java/org/apache/helix/controller/changedetector/TestResourceChangeDetector.java:435]: nit: Can we do 
Assert.assertTrue(_clusterVerifier.verify()); ?

**Review by jiajunwang — APPROVED** [review]
Please address the other comments.

File: helix-core/src/test/java/org/apache/helix/controller/changedetector/TestResourceChangeDetector.java (hunk 5 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/controller/changedetector/TestResourceChangeDetector.java:435]: Changed.

**Comment by kaisun2000** [comment]
This PR is approved, please help to merge in.

> Fix several tests with resource leakage

**Pull request merged**